### PR TITLE
Add SSL support

### LINKED ISSUE #3: Add SSL support
The URI scheme supports only TCP, no SSL (the author doesn't need it because the client is run in "safe" production environments). For the async client, it should be straightforward to support SSL by means of the Endpoint API. Contributions are welcome!

### PR DESCRIPTION
**Pull request by joshainglis**
Resolves #3 
@nikipore: Let me know if you have any concerns.

### CODE REVIEW

**Comment by nikipore** [comment]
Thank you very much for your contribution. At a first glance, it looks clean, but I'll have to take a second and more thorough look and will come back to you. For the time being, would you please add test cases?

**Comment by joshainglis** [comment]
Thanks @nikipore :) I'll get the tests written and pushed up shortly

**Comment by nikipore** [comment]
Your code is very nice and concise, and surprisingly straightforward. Some (hopefully) minor comments:

1. Please remove the annotation comments which suppress warnings in your particular IDE or linter.
2. Please adapt to the naming conventions in this project which aren't entirely PEP 8. To be specific: variables and function parameters are mixedCase not lower_case_with_underscores.
3. Please replace the default value of `ciphers` by a constant (defined in the `ssl` module?).
4. Please add examples for the usage of both `sync` and `async` client. This would also make it easier to understand to SSL noobs (like myself) why the `async` code supports SSL without modifications; it apparently works via the URI and `twisted.internet.endpoints.clientFromString`, but it doesn't seem to make use of `stompest.config.SSLConfig`, so the latter seems to be specific to the `sync` client and should therefore reside in a module inside `stompest.sync` not `stompest`. It would be interesting for SSL noobs (like myself) to understand inhowfar the `sync` and the `async` versions differ in terms flexibility of SSL parameters.

**Comment by joshainglis** [comment]
Thanks for the feedback @nikipore!
1 and 2 are done. As for 3 and 4, I'm going to add a bit more code to ensure the `SSLConfig` is used consistently for both `sync` and `async` connections. That way SSL usage for `sync` and `async` should be consistent.
Working on this currently.

File: src/async/stompest/async/client.py [hunk 1]
```diff
@@ -55,11 +54,22 @@ class Stomp(object):
...
 
     def __init__(self, config, listenersFactory=None, endpointFactory=None):
+        """
+
+        :param config:
+        :type config: stompest.config.StompConfig
+        :param listenersFactory:
+        :type listenersFactory:
+        :param endpointFactory:
+        :type endpointFactory: (dict[str, str], stompest.config.StompConfig) -> twisted.internet.endpoints.SSL4ClientEndpoint | twisted.internet.endpoints.TCP4ClientEndpoint
```
Comment (nikipore) [inline src/async/stompest/async/client.py]: The first parameter is a `stompest.protocol.broker.Broker` not a `dict`.

File: src/async/stompest/async/sslContext.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
+from _ssl import CERT_NONE, CERT_OPTIONAL, CERT_REQUIRED
```
Comment (nikipore) [inline src/async/stompest/async/sslContext.py]: No `mixedCase`for module names. Use `sslcontext`or just `ssl`.

File: src/async/stompest/async/sslContext.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    def getContext(self):
+        """
+        :return: an OpenSSL context used by Twisted
+        :rtype: OpenSSL.SSL.Context
+        """
+        try:
+            ctx = self._contextFactory(self.method)
+        except AttributeError:
+            raise ValueError(
+                "Invalid sslVersion (method). Valid versions are {}".format(self.validSSLVersionStrings)
```
Comment (nikipore) [inline src/async/stompest/async/sslContext.py]: Use apostroph not quotes as string delimiters.

File: src/core/stompest/sync/transport.py [hunk 4]
```diff
@@ -45,6 +60,13 @@ def connect(self, timeout=None):
         kwargs = {} if (timeout is None) else {'timeout': timeout}
         try:
             self._socket = socket.create_connection((self.host, self.port), **kwargs)
+            if self.protocol == 'ssl':
```
Comment (nikipore) [inline src/core/stompest/sync/transport.py]: Maybe extract a socket wrapper method.

File: src/async/stompest/async/util.py [hunk 5]
```diff
@@ -87,10 +91,48 @@ def _reload(result=None):
...
+    :param timeout:
+    :type timeout: int
+    :param broker:
+    :type broker: stompest.protocol.broker.Broker
+    :param config:
+    :type config: stompest.config.StompConfig
+    :return:
+    :rtype: twisted.internet.endpoints.TCP4ClientEndpoint | twisted.internet.endpoints.SSL4ClientEndpoint
+    """
+    protocol = broker['protocol'].lower()
```
Comment (nikipore) [inline src/async/stompest/async/util.py]: Replace with `broker.protocol`.

File: src/core/stompest/protocol/broker.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    def fromUri(cls, uri):
+        """
+        :param uri:
+        :type uri: str
+        :return:
+        :rtype: Broker
+        """
+        return cls(**cls._REGEX_URI.match(uri).groupdict())
+
+    def __getitem__(self, item):
```
Comment (nikipore) [inline src/core/stompest/protocol/broker.py]: Backwards compatibility (`dict`like behavior of `Broker`) really required?

File: src/core/stompest/config/__init__.py [hunk 7]
```diff
@@ -1,3 +1,63 @@
...
+        SSL options for both sync and async connections.
+        Defaults should work fine for most situations.
+        Set verifyCertificate=False if you do not wish to verify certs.
+        Typicall
+        Only change if you know what you're doing.
+
+        :param verifyCertificate: Valid inputs are: True (CERT_REQUIRED), False (CERT_NONE)
+                                  Defaults to True.
+        :type verifyCertificate: bool
+        :param sslVersion: Valid inputs: 'SSLv2', 'SSLv23', 'SSLv3', 'TLSv1', 'TLSv1_1', 'TLSv1_2'
```
Comment (nikipore) [inline src/core/stompest/config/__init__.py]: If possible use constants not strings.

File: src/core/stompest/config/__init__.py [hunk 8]
```diff
@@ -1,3 +1,63 @@
+from _ssl import CERT_REQUIRED, CERT_NONE
+from ssl import _DEFAULT_CIPHERS
+
+
+class SSLConfig(object):
+    defaultSSLOptions = [
```
Comment (nikipore) [inline src/core/stompest/config/__init__.py]: If possible use constants not strings.

File: src/core/stompest/config/__init__.py [hunk 9]
```diff
@@ -1,3 +1,63 @@
...
+        :param verifyCertificate: Valid inputs are: True (CERT_REQUIRED), False (CERT_NONE)
+                                  Defaults to True.
+        :type verifyCertificate: bool
+        :param sslVersion: Valid inputs: 'SSLv2', 'SSLv23', 'SSLv3', 'TLSv1', 'TLSv1_1', 'TLSv1_2'
+                            defaults to 'SSLv23'
+        :type sslVersion: int
+        :param options: SSL options. Warnings will raise for invalid options.
+                        defaults to ['OP_NO_SSLv2', 'OP_NO_SSLv3', 'OP_NO_COMPRESSION', 'OP_CIPHER_SERVER_PREFERENCE', 'OP_SINGLE_DH_USE', 'OP_SINGLE_ECDH_USE']
+        :type options: list[str]
+        :param ciphers: defaults to 'ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+HIGH:DH+HIGH:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+HIGH:RSA+3DES:!aNULL:!eNULL:!MD5'
```
Comment (nikipore) [inline src/core/stompest/config/__init__.py]: If possible use constants not strings.

File: src/core/stompest/config/__init__.py [hunk 10]
```diff
@@ -1,3 +1,63 @@
+from _ssl import CERT_REQUIRED, CERT_NONE
+from ssl import _DEFAULT_CIPHERS
+
+
+class SSLConfig(object):
```
Comment (nikipore) [inline src/core/stompest/config/__init__.py]: It doesn't feel right that SSL and non-SSL stuff is mixed in this module. But maybe I'm wrong.

**Comment by nikipore** [comment]
Great work! I added some minor inline notes. When these are adressed and an example for SSL usage is added, I will accept the pull request, test it myself and build a new release.

File: src/core/stompest/protocol/broker.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import re
+
+
+class Broker(object):
+    PROTOCOL_TCP = 'tcp'
+    PROTOCOL_SSL = 'ssl'
+    PROTOCOLS = {PROTOCOL_TCP, PROTOCOL_SSL}
```
Comment (nikipore) [inline src/core/stompest/protocol/broker.py]: Use `set(...)` not `{...}` because `stompest` needs to remain backward compatible to Python 2.6.

**Comment by nikipore** [comment]
Very nice indeed.

I pulled the code and tried to test it. It doesn't run with Python 2.6, but it has to. The reason is that my employer has to use it in a context which is restricted to Python 2.6 for a few months to go. Problems I ran into (before I stopped):
* Use `set(...)` not `{...}`.
* The name `ssl._DEFAULT_CIPHERS` doesn't exist.
* A `.format(...)` method is broken.

Please retest with Python 2.6. Oh, and the docstring for `stompest.config.sslconfig.SSLConfig` still refers to strings not constants.

**Comment by joshainglis** [comment]
Got the low-hanging fruit there, but man, SSL on Python 2.6 is a pain haha!
I'm a tad time-poor at the moment, but I will endeavour to finish the 2.6 support as soon as I can.

**Comment by nikipore** [comment]
How much pain is it? Will the functionality be restricted. The one important user (my employer) I know of who still relies on Python 2.6 compatibility will drop that requirement in a few weeks. So, an alternative would be that I make a poll on the stompest Google group.

**Comment by nikipore** [comment]
I just retested. The unit tests still fail (only tried the `stompest.core` tests), the overall reason being that `ssl.create_default_context()` is not defined, leading to an `AttributeError: 'module' object has no attribute 'create_default_context'`.

**Comment by nikipore** [comment]
Hmm ... I just realized that the original tests do pass that you haven't added any test cases for SSL connections yet. And the examples and documentation of the relevant modules are still missing. Would you please be so kind to add these before I build a version 2.2.0?

**Comment by joshainglis** [comment]
Hi @nikipore, thanks for the accepted pull.
I'll endeavour to get some tests and examples added this coming weekend.

**Pull request merged**